cryptographic_estimators/UOVEstimator/uov_problem.py

```diff
@@ -0,0 +1,188 @@
+# ****************************************************************************
+# Copyright 2023 Technology Innovation Institute
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# ****************************************************************************
+
+
+from ..base_problem import BaseProblem
+from sage.arith.misc import is_prime_power
+from ..MQEstimator.mq_helper import ngates
+from .uov_constants import *
+from math import log2
+from sage.functions.other import ceil
+
+class UOVProblem(BaseProblem):
+    """
+    Construct an instance of UOVProblem.
+
+    INPUT: 
+        
+    - ``n`` -- number of variables
+    - ``m`` -- number of polynomials
+    - ``q`` -- order of the finite field
+    - ``theta`` -- exponent of the conversion factor (default: 2)
+    - If ``0 <= theta <= 2``, every multiplication in GF(q) is counted as `log2(q) ^ theta` binary operation.
+    - If ``theta = None``, every multiplication in GF(q) is counted as `2 * log2(q) ^ 2 + log2(q)` binary operation.
+    - ``memory_bound`` -- maximum allowed memory to use for solving the problem (default: inf)
+
+    """
+
+    def __init__(self, n: int, m: int, q:int, **kwargs):
+        super().__init__(**kwargs)
+
+        theta = kwargs.get("theta", 2)
+        cost_one_hash = kwargs.get("cost_one_hash", 17)
+        
+        if n < 1:
+            raise ValueError("n must be >= 1")
+
+        if m < 1:
+            raise ValueError("m must be >= 1")
+
+        if n <= m:
+            raise ValueError("n must be > m")
+
+        if not is_prime_power(q):
+            raise ValueError("q must be a prime power")
+        
+        if theta is not None and not (0 <= theta <= 2):
+            raise ValueError("theta must be either None or 0 <=theta <= 2")
+
+        if cost_one_hash < 0:
+            raise ValueError("The cost of computing one hash must be >= 0")
+  
+        self.parameters[UOV_NUMBER_VARIABLES] = n
+        self.parameters[UOV_NUMBER_POLYNOMIALS] = m
+        self.parameters[UOV_FIELD_SIZE] = q
+        self._theta = theta
+        self._cost_one_hash = cost_one_hash
```

Comment:
is this cost in bit or in basic operations?

Reply:
Cost in bit operations. We have a function `hashes_to_basic_operations` to transforme hashes into basic operations

Reply:
`is this cost in bit or in basic operations?` clariification added to the doc.